http/src/v0/block.rs

```diff
@@ -102,8 +81,85 @@ pub fn put<T: IpfsTypes>(
     path!("block" / "put")
         .and(with_ipfs(ipfs))
         .and(query::<PutQuery>())
-        .and(multipart::form())
-        .and_then(put_query)
+        .and(warp::header::<Mime>("content-type"))
+        .and(warp::body::stream())
+        .and_then(inner_put)
+}
+
+async fn inner_put<T: IpfsTypes>(
+    ipfs: Ipfs<T>,
+    opts: PutQuery,
+    mime: Mime,
+    body: impl Stream<Item = Result<impl Buf, warp::Error>> + Unpin,
+) -> Result<impl Reply, Rejection> {
+    let boundary = mime.get_param("boundary").map(|v| v.to_string()).unwrap();
+    let mut stream = MultipartStream::new(boundary, body.map_ok(|mut buf| buf.to_bytes()));
+
+    // store the first good field here
+    let mut buffer = Vec::new();
+
+    loop {
+        match stream.try_next().await {
+            Ok(Some(mut field)) => {
+                // [ipfs http api] says we should expect a "data" but examples use "file" as the
+                // form field name. conformance tests also use latter.
+                //
+                // [ipfs http api]: https://docs.ipfs.io/reference/http/api/#api-v0-block-put
+
+                let name = field
+                    .name()
+                    .map_err(|_| StringError::from("invalid field name"))?;
+
+                if name != "data" {
+                    log::warn!("block/put: discarding part with name {:?}", name);
+                    continue;
+                }
+
+                if !buffer.is_empty() {
+                    return Err(StringError::from(
+                        "return error for multiple blocks (this API should only accept one)",
+                    )
+                    .into());
+                }
+
+                loop {
+                    let next = field.try_next().await.map_err(|e| {
+                        StringError::from(format!("IO error while reading field bytes: {}", e))
+                    })?;
+
+                    match next {
+                        Some(bytes) => buffer.extend_from_slice(bytes.as_ref()),
+                        None => break,
+                    }
+                }
+            }
+            Ok(None) => break,
+            Err(e) => {
+                todo!("IO error while reading next field: {}", e);
+            }
+        }
+    }
+
+    if buffer.is_empty() {
+        return Err(StringError::from("no suitable chunk found").into());
+    }
+
+    let data = buffer.into_boxed_slice();
+
+    let digest = opts.digest()?(&data);
```

Comment:
Interesting that rustfmt doesn't put parens around this `(opts.digest()?)(&data)` and rustc doesn't recommend those, clippy is silent. Can't say I would like to go searching if an issue has already been created on this so I assume this has already been bikeshed.